README.md
# Checksum Finder

Checksum Finder is an application to test file integrity by means of the MD5 hash algorithm. SHA-1 support is to be implemented soon.

## Usage

To test a file, simply press the Browse button (or Ctrl+O) and find the file that you wish to find the checksum of. The MD5 will automatically be calculated and displayed in the designated area- for large files (several hundred megabytes), give it a few seconds to process the file.

## Build Instructions

Running this will install all of the needed build tools for this application:

    sudo apt-get install build-essential libwxgtk2.8-dev

Checksum Finder is written in C++ using wxWidgets 2.8.12. If your compiler is having trouble finding the wxWidgets library and you know that you have it installed, check that there is output for the following commands:

    wx-config --cppflags
    wx-config --libs

Run ```make``` to build the program, or ```make rebuild``` to completely rebuild the program and all object files.

```make clean``` will delete all object files while leaving the application intact, and ```make delete``` or ```make erase``` will completely delete all build artifacts, including the compiled application.


CHANGELOG.md
# Checksum Finder Change Log

All changes to Checksum Finder will be logged in this file with each release.

## 0.1

This is the initial alpha release of Checksum Finder.

### Goals for Next Release

 + Have the File and Compare text controls accept Enter.
 + Correct any bugs discovered and get user feedback from this alpha.

#### Patches

 + 0.1.0.1 - Fixed debian build.


src/Program.cpp
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file Program.cpp - Main entry point for the application.
 * Shows an instance of frmMain and enters the main event loop.
**/

#include "Program.h"
#include "frmMain.h"

IMPLEMENT_APP(Sandbox);

bool Sandbox::OnInit()
{
    frmMain *main = new frmMain();
    main->Show();
    SetTopWindow(main);

    return true;
}



src/Include/MD5.h
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file MD5.h - Header file for the MD5 class and MD5-related functions.
**/

#ifndef MD5_H_INCLUDED
#define MD5_H_INCLUDED

#include <ostream>
#include <cstdio>
#include <cstring>
#include <wx/string.h>

using namespace std;

class MD5
{
public:
    typedef unsigned int size_type; // Must be 32-bit

    MD5();
    MD5(const string &text);
    void update(const unsigned char *buf, size_type length);
    void update(const char *buf, size_type length);
    MD5 &finalize();
    string hexdigest() const;
    friend ostream& operator<<(ostream&, MD5 md5);

private:
    void init();
    typedef unsigned char uint1; // 8-bit
    typedef unsigned int uint4; // 32-bit
    enum { blocksize = 64 };

    void transform(const uint1 block[blocksize]);
    static void decode(uint4 output[], const uint1 input[], size_type len);
    static void encode(uint1 output[], const uint4 input[], size_type len);

    bool finalized;
    uint1 buffer[blocksize]; // Bytes that didn't fit in last 64 byte chunk
    uint4 count[2]; // 64-bit counter for number of bits (lo, hi)
    uint4 state[4]; // Digest so far
    uint1 digest[16]; // The result

    // Low level logic operations
    static inline uint4 F(uint4 x, uint4 y, uint4 z);
    static inline uint4 G(uint4 x, uint4 y, uint4 z);
    static inline uint4 H(uint4 x, uint4 y, uint4 z);
    static inline uint4 I(uint4 x, uint4 y, uint4 z);
    static inline uint4 rotate_left(uint4 x, int n);
    static inline void FF(uint4 &a, uint4 b, uint4 c, uint4 d,
                          uint4 x, uint4 s, uint4 ac);
    static inline void GG(uint4 &a, uint4 b, uint4 c, uint4 d,
                          uint4 x, uint4 s, uint4 ac);
    static inline void HH(uint4 &a, uint4 b, uint4 c, uint4 d,
                          uint4 x, uint4 s, uint4 ac);
    static inline void II(uint4 &a, uint4 b, uint4 c, uint4 d,
                          uint4 x, uint4 s, uint4 ac);
};

wxString encryptMD5(wxString str);

#endif // MD5_H_INCLUDED



src/Classes/MD5.cpp
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file MD5.cpp - Class for MD5 encryption and MD5-related functions.
**/

#include "MD5.h"

// Constants for MD5Transform routine.
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

// F, G, H and I are basic MD5 functions.
inline MD5::uint4 MD5::F(uint4 x, uint4 y, uint4 z)
{
    return (x & y) | (~x & z);
}

inline MD5::uint4 MD5::G(uint4 x, uint4 y, uint4 z)
{
    return (x & z) | (y & ~z);
}

inline MD5::uint4 MD5::H(uint4 x, uint4 y, uint4 z)
{
    return x ^ y ^ z;
}

inline MD5::uint4 MD5::I(uint4 x, uint4 y, uint4 z)
{
    return y ^ (x | ~z);
}

// rotate_left rotates x left n bits.
inline MD5::uint4 MD5::rotate_left(uint4 x, int n)
{
    return (x << n) | (x >> (32 - n));
}

// FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
// Rotation is separate from addition to prevent recomputation.
inline void MD5::FF(uint4 &a, uint4 b, uint4 c, uint4 d,
                    uint4 x, uint4 s, uint4 ac)
{
    a = rotate_left(a + F(b, c, d) + x + ac, s) + b;
}
inline void MD5::GG(uint4 &a, uint4 b, uint4 c, uint4 d,
                    uint4 x, uint4 s, uint4 ac)
{
    a = rotate_left(a + G(b, c, d) + x + ac, s) + b;
}
inline void MD5::HH(uint4 &a, uint4 b, uint4 c, uint4 d,
                    uint4 x, uint4 s, uint4 ac)
{
    a = rotate_left(a + H(b, c, d) + x + ac, s) + b;
}
inline void MD5::II(uint4 &a, uint4 b, uint4 c, uint4 d,
                    uint4 x, uint4 s, uint4 ac)
{
    a = rotate_left(a + I(b, c, d) + x + ac, s) + b;
}

// Default constructor, just initailize
MD5::MD5()
{
    init();
}

// Nifty shortcut constructor, compute MD5 for string, finalize it right away
MD5::MD5(const string &text)
{
    init();
    update(text.c_str(), text.length());
    finalize();
}

void MD5::init()
{
    finalized = false;

    count[0] = 0;
    count[1] = 0;

    // Load magic initialization constants.
    state[0] = 0x67452301;
    state[1] = 0xefcdab89;
    state[2] = 0x98badcfe;
    state[3] = 0x10325476;
}

// Decodes input (unsigned char) into output (uint4).
// Assumes len is a multiple of 4.
void MD5::decode(uint4 output[], const uint1 input[], size_type len)
{
    for (unsigned int i = 0, j = 0; j < len; ++i, j += 4)
        output[i] = ((uint4)input[j]) | (((uint4)input[j + 1]) << 8) |
                    (((uint4)input[j + 2]) << 16) |
                    (((uint4)input[j + 3]) << 24);
}

// Encodes input (uint4) into output (unsigned char).
// Assumes len is a multiple of 4.
void MD5::encode(uint1 output[], const uint4 input[], size_type len)
{
    for (size_type i = 0, j = 0; j < len; ++i, j += 4)
    {
        output[j] = input[i] & 0xff;
        output[j + 1] = (input[i] >> 8) & 0xff;
        output[j + 2] = (input[i] >> 16) & 0xff;
        output[j + 3] = (input[i] >> 24) & 0xff;
    }
}

// Apply MD5 algorithm on a block
void MD5::transform(const uint1 block[blocksize])
{
    uint4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];
    decode(x, block, blocksize);

    // Round 1
    FF(a, b, c, d, x[ 0], S11, 0xd76aa478); // 1
    FF(d, a, b, c, x[ 1], S12, 0xe8c7b756); // 2
    FF(c, d, a, b, x[ 2], S13, 0x242070db); // 3
    FF(b, c, d, a, x[ 3], S14, 0xc1bdceee); // 4
    FF(a, b, c, d, x[ 4], S11, 0xf57c0faf); // 5
    FF(d, a, b, c, x[ 5], S12, 0x4787c62a); // 6
    FF(c, d, a, b, x[ 6], S13, 0xa8304613); // 7
    FF(b, c, d, a, x[ 7], S14, 0xfd469501); // 8
    FF(a, b, c, d, x[ 8], S11, 0x698098d8); // 9
    FF(d, a, b, c, x[ 9], S12, 0x8b44f7af); // 10
    FF(c, d, a, b, x[10], S13, 0xffff5bb1); // 11
    FF(b, c, d, a, x[11], S14, 0x895cd7be); // 12
    FF(a, b, c, d, x[12], S11, 0x6b901122); // 13
    FF(d, a, b, c, x[13], S12, 0xfd987193); // 14
    FF(c, d, a, b, x[14], S13, 0xa679438e); // 15
    FF(b, c, d, a, x[15], S14, 0x49b40821); // 16

    // Round 2
    GG(a, b, c, d, x[ 1], S21, 0xf61e2562); // 17
    GG(d, a, b, c, x[ 6], S22, 0xc040b340); // 18
    GG(c, d, a, b, x[11], S23, 0x265e5a51); // 19
    GG(b, c, d, a, x[ 0], S24, 0xe9b6c7aa); // 20
    GG(a, b, c, d, x[ 5], S21, 0xd62f105d); // 21
    GG(d, a, b, c, x[10], S22,  0x2441453); // 22
    GG(c, d, a, b, x[15], S23, 0xd8a1e681); // 23
    GG(b, c, d, a, x[ 4], S24, 0xe7d3fbc8); // 24
    GG(a, b, c, d, x[ 9], S21, 0x21e1cde6); // 25
    GG(d, a, b, c, x[14], S22, 0xc33707d6); // 26
    GG(c, d, a, b, x[ 3], S23, 0xf4d50d87); // 27
    GG(b, c, d, a, x[ 8], S24, 0x455a14ed); // 28
    GG(a, b, c, d, x[13], S21, 0xa9e3e905); // 29
    GG(d, a, b, c, x[ 2], S22, 0xfcefa3f8); // 30
    GG(c, d, a, b, x[ 7], S23, 0x676f02d9); // 31
    GG(b, c, d, a, x[12], S24, 0x8d2a4c8a); // 32

    // Round 3
    HH(a, b, c, d, x[ 5], S31, 0xfffa3942); // 33
    HH(d, a, b, c, x[ 8], S32, 0x8771f681); // 34
    HH(c, d, a, b, x[11], S33, 0x6d9d6122); // 35
    HH(b, c, d, a, x[14], S34, 0xfde5380c); // 36
    HH(a, b, c, d, x[ 1], S31, 0xa4beea44); // 37
    HH(d, a, b, c, x[ 4], S32, 0x4bdecfa9); // 38
    HH(c, d, a, b, x[ 7], S33, 0xf6bb4b60); // 39
    HH(b, c, d, a, x[10], S34, 0xbebfbc70); // 40
    HH(a, b, c, d, x[13], S31, 0x289b7ec6); // 41
    HH(d, a, b, c, x[ 0], S32, 0xeaa127fa); // 42
    HH(c, d, a, b, x[ 3], S33, 0xd4ef3085); // 43
    HH(b, c, d, a, x[ 6], S34,  0x4881d05); // 44
    HH(a, b, c, d, x[ 9], S31, 0xd9d4d039); // 45
    HH(d, a, b, c, x[12], S32, 0xe6db99e5); // 46
    HH(c, d, a, b, x[15], S33, 0x1fa27cf8); // 47
    HH(b, c, d, a, x[ 2], S34, 0xc4ac5665); // 48

    // Round 4
    II(a, b, c, d, x[ 0], S41, 0xf4292244); // 49
    II(d, a, b, c, x[ 7], S42, 0x432aff97); // 50
    II(c, d, a, b, x[14], S43, 0xab9423a7); // 51
    II(b, c, d, a, x[ 5], S44, 0xfc93a039); // 52
    II(a, b, c, d, x[12], S41, 0x655b59c3); // 53
    II(d, a, b, c, x[ 3], S42, 0x8f0ccc92); // 54
    II(c, d, a, b, x[10], S43, 0xffeff47d); // 55
    II(b, c, d, a, x[ 1], S44, 0x85845dd1); // 56
    II(a, b, c, d, x[ 8], S41, 0x6fa87e4f); // 57
    II(d, a, b, c, x[15], S42, 0xfe2ce6e0); // 58
    II(c, d, a, b, x[ 6], S43, 0xa3014314); // 59
    II(b, c, d, a, x[13], S44, 0x4e0811a1); // 60
    II(a, b, c, d, x[ 4], S41, 0xf7537e82); // 61
    II(d, a, b, c, x[11], S42, 0xbd3af235); // 62
    II(c, d, a, b, x[ 2], S43, 0x2ad7d2bb); // 63
    II(b, c, d, a, x[ 9], S44, 0xeb86d391); // 64

    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;

    // Zeroize sensitive information.
    memset(x, 0, sizeof x);
}

// MD5 block update operation.
// Continues an MD5 message-digest operation, processing another message block
void MD5::update(const unsigned char input[], size_type length)
{
    // Compute number of bytes mod 64
    size_type index = count[0] / 8 % blocksize;

    // Update number of bits
    if ((count[0] += (length << 3)) < (length << 3))
        count[1]++;

    count[1] += (length >> 29);

    // Number of bytes we need to fill in buffer
    size_type firstpart = 64 - index;

    size_type i;

    // Transform as many times as possible.
    if (length >= firstpart)
    {
        // Fill buffer first, transform
        memcpy(&buffer[index], input, firstpart);
        transform(buffer);

        // Transform chunks of blocksize (64 bytes)
        for (i = firstpart; i + blocksize <= length; i += blocksize)
            transform(&input[i]);

        index = 0;
    }
    else i = 0;

    // Buffer remaining input
    memcpy(&buffer[index], &input[i], length - i);
}
// For convenience provide an overload with signed char
void MD5::update(const char input[], size_type length)
{
    update((const unsigned char*)input, length);
}

// MD5 finalization. Ends an MD5 message-digest operation, writing the
// the message digest and zeroizing the context.
MD5& MD5::finalize()
{
    static unsigned char padding[64] =
    {
        0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };

    if (!finalized)
    {
        // Save number of bits
        unsigned char bits[8];
        encode(bits, count, 8);

        // Pad out to 56 mod 64.
        size_type index = count[0] / 8 % 64;
        size_type padLen = (index < 56) ? (56 - index) : (120 - index);
        update(padding, padLen);

        // Append length (before padding)
        update(bits, 8);

        // Store state in digest
        encode(digest, state, 16);

        // Zeroize sensitive information.
        memset(buffer, 0, sizeof buffer);
        memset(count, 0, sizeof count);

        finalized = true;
    }

    return *this;
}

// Return hex representation of digest as string
string MD5::hexdigest() const
{
    if (!finalized)
        return "";

    char buf[33];

    for (int i = 0; i < 16; i++)
        sprintf(buf + i * 2, "%02x", digest[i]);

    buf[32] = 0;

    return string(buf);
}

ostream& operator<<(ostream& out, MD5 md5)
{
    return out << md5.hexdigest();
}

wxString encryptMD5(wxString str)
{
    return wxString((new MD5(string(str.mb_str())))->hexdigest().c_str(),
                    wxConvUTF8);
}



src/Include/Program.h
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file Program.h - Header file for the program's main entry point.
**/

#ifndef PROGRAM_H_INCLUDED
#define PROGRAM_H_INCLUDED

#include <wx/wx.h>

class Sandbox : public wxApp
{
public:
    virtual bool OnInit();
};

#endif // PROGRAM_H_INCLUDED



src/Include/frmMain.h
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file frmMain.h - Header file for the frmMain class.
**/

#ifndef FRMMAIN_H_INCLUDED
#define FRMMAIN_H_INCLUDED

#include <wx/wx.h>
#include <wx/aboutdlg.h>
#include <wx/colour.h>
#include <wx/ffile.h>
#include <wx/filedlg.h>
#include <wx/menu.h>
#include <wx/string.h>
#include "Clipboard.h"
#include "MD5.h"

class frmMain : public wxFrame
{
public:

    frmMain(void);

    /** Menu Bar **/

    wxMenuBar *menuBar;

    wxMenu *mnuFile;
    //
    wxMenuItem *miOpen;
    // Separator
    wxMenuItem *miQuit;

    wxMenu *mnuHelp;
    //
    wxMenuItem *miAbout;

    /** Main Window **/

    wxPanel *pnlMargin;
    wxBoxSizer *vboxMargin;

    wxPanel *pnlMain;
    wxBoxSizer *vboxMain;

    wxBoxSizer *hboxFileInput;
    wxStaticText *lblFile;
    wxTextCtrl *txtFile;
    wxButton *btnBrowse;

    wxBoxSizer *hboxMD5Output;
    wxStaticText *lblMD5;
    wxTextCtrl *txtMD5;
    wxButton *btnCopyMD5;

    // :TODO: 2012-05-27 gbchaosmaster - Implement SHA-1.

//  wxBoxSizer *hboxSHA1Output;
//  wxStaticText *lblSHA1;
//  wxTextCtrl *txtSHA1;
//  wxButton *btnCopySHA1;

    wxBoxSizer *hboxCompare;
    wxStaticText *lblCompare;
    wxTextCtrl *txtCompare;
    wxButton *btnCompare;

    /** Events **/

    void selectFile(wxCommandEvent &event);
    void compare(wxCommandEvent &event);
    void copyMD5(wxCommandEvent &event);
//  void copySHA1(wxCommandEvent &event);
    void showAbout(wxCommandEvent &event);
    void quit(wxCommandEvent &event);

    /** Functions **/

    void copyChecksum(wxString str);

private:

    /** Identifiers **/

    static const long ID_COMPARE;
    static const long ID_COPY_MD5;
//  static const long ID_COPY_SHA1;
    static const long ID_BROWSE;
    static const long ID_OPEN;
};

#endif // FRMMAIN_H_INCLUDED



src/Include/Clipboard.h
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file Clipboard.h - Header file for the Clipboard class.
**/

#ifndef CLIPBOARD_H_INCLUDED
#define CLIPBOARD_H_INLCUDED

#include <wx/wx.h>
#include <wx/clipbrd.h>
#include <wx/string.h>

class Clipboard
{
public:
    static void SetClipboard(wxString str);
    static wxString GetClipboard(void);
};

#endif // CLIPBOARD_H_INLCUDED



src/Classes/Clipboard.cpp
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file Clipboard.cpp - Class with static methods for clipboard handling.
**/

#include "Clipboard.h"

void Clipboard::SetClipboard(wxString str)
{
    if (wxTheClipboard->Open())
    {
        wxTheClipboard->SetData(new wxTextDataObject(str));
        wxTheClipboard->Close();
    }
}
wxString Clipboard::GetClipboard(void)
{
    wxString str = _T("");
    if (wxTheClipboard->Open())
    {
        if (wxTheClipboard->IsSupported(wxDF_TEXT))
        {
            wxTextDataObject data;
            wxTheClipboard->GetData(data);
            str = data.GetText();
        }
        wxTheClipboard->Close();
    }
    return str;
}



src/Frames/frmMain.cpp
/**
 * A program for verifying file integrity.
 * Copyright (C) 2012 Vinny Diehl
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

/**
 * @file frmMain.cpp - The main wxFrame for the application.
**/

#include "frmMain.h"

const wxString TITLE = _T("Checksum Finder");
const int MIN_WIDTH = 550;
const char MARGIN = 5;
const char SEPARATOR = 2;

/** Initialize Identifiers **/

const long frmMain::ID_COMPARE = wxNewId();
const long frmMain::ID_COPY_MD5 = wxNewId();
//const long frmMain::ID_COPY_SHA1 = wxNewId();
const long frmMain::ID_OPEN = wxNewId();
const long frmMain::ID_BROWSE = wxNewId();

frmMain::frmMain(void) : wxFrame(NULL, wxID_ANY, TITLE)
{
    /** Menu Bar **/

    menuBar = new wxMenuBar;

    // File
    mnuFile = new wxMenu;

    miOpen = new wxMenuItem(mnuFile, ID_OPEN, _T("&Open\tCtrl+O"));
    miQuit = new wxMenuItem(mnuFile, wxID_EXIT, _T("&Quit\tCtrl+W"));

    mnuFile->Append(miOpen);
    mnuFile->AppendSeparator();
    mnuFile->Append(miQuit);

    // Help
    mnuHelp = new wxMenu;

    miAbout = new wxMenuItem(mnuHelp, wxID_ABOUT, _T("&About\tF1"));

    mnuHelp->Append(miAbout);

    // Assemble the menu bar
    menuBar->Append(mnuFile, _T("&File"));
    menuBar->Append(mnuHelp, _T("&Help"));
    SetMenuBar(menuBar);

    // Connect menu bar events
    Connect(ID_OPEN, wxEVT_COMMAND_MENU_SELECTED,
            wxCommandEventHandler(frmMain::selectFile));
    Connect(wxID_EXIT, wxEVT_COMMAND_MENU_SELECTED,
            wxCommandEventHandler(frmMain::quit));
    Connect(wxID_ABOUT, wxEVT_COMMAND_MENU_SELECTED,
            wxCommandEventHandler(frmMain::showAbout));

    /** Main Window **/

    pnlMargin = new wxPanel(this, wxID_ANY);
    vboxMargin = new wxBoxSizer(wxVERTICAL);

    pnlMain = new wxPanel(pnlMargin, wxID_ANY);
    vboxMain = new wxBoxSizer(wxVERTICAL);

    hboxFileInput = new wxBoxSizer(wxHORIZONTAL);

    lblFile = new wxStaticText(pnlMain, wxID_ANY, _T("File:"));
    txtFile = new wxTextCtrl(pnlMain, wxID_ANY);
    btnBrowse = new wxButton(pnlMain, ID_BROWSE, _T("&Browse..."));

    hboxFileInput->Add(lblFile, 0, wxALIGN_CENTER_VERTICAL | wxRIGHT, 5);
    hboxFileInput->Add(txtFile, 1, wxALIGN_CENTER_VERTICAL);
    hboxFileInput->Add(btnBrowse, 0, wxALIGN_CENTER_VERTICAL | wxLEFT, 5);

    vboxMain->Add(hboxFileInput, 0, wxEXPAND);
    vboxMain->Add(-1, SEPARATOR);

    // End hboxFileInput

    hboxMD5Output = new wxBoxSizer(wxHORIZONTAL);

    lblMD5 = new wxStaticText(pnlMain, wxID_ANY, _T("MD5 Checksum:"));
    txtMD5 = new wxTextCtrl(pnlMain, wxID_ANY, wxEmptyString,
                            wxDefaultPosition, wxDefaultSize, wxTE_READONLY);
    btnCopyMD5 = new wxButton(pnlMain, ID_COPY_MD5, _T("Copy &MD5"));

    txtMD5->SetBackgroundColour(*wxLIGHT_GREY);

    hboxMD5Output->Add(lblMD5, 0, wxALIGN_CENTER_VERTICAL | wxRIGHT, 5);
    hboxMD5Output->Add(txtMD5, 1, wxALIGN_CENTER_VERTICAL);
    hboxMD5Output->Add(btnCopyMD5, 0, wxALIGN_CENTER_VERTICAL | wxLEFT, 5);

    vboxMain->Add(hboxMD5Output, 0, wxEXPAND);
    vboxMain->Add(-1, SEPARATOR);

    // End hboxMD5Output

    // :TODO: 2012-05-27 gbchaosmaster - Implement SHA-1.

//  hboxSHA1Output = new wxBoxSizer(wxHORIZONTAL);

//  lblSHA1 = new wxStaticText(pnlMain, wxID_ANY, _T("SHA1 Checksum:"));
//  txtSHA1 = new wxTextCtrl(pnlMain, wxID_ANY, wxEmptyString,
//                           wxDefaultPosition, wxDefaultSize, wxTE_READONLY);
//  btnCopySHA1 = new wxButton(pnlMain, ID_COPY_SHA1, _T("Copy &SHA1"));

//  txtSHA1->SetBackgroundColour(*wxLIGHT_GREY);

//  hboxSHA1Output->Add(lblSHA1, 0, wxALIGN_CENTER_VERTICAL | wxRIGHT, 5);
//  hboxSHA1Output->Add(txtSHA1, 1, wxALIGN_CENTER_VERTICAL);
//  hboxSHA1Output->Add(btnCopySHA1, 0, wxALIGN_CENTER_VERTICAL | wxLEFT, 5);

//  vboxMain->Add(hboxSHA1Output, 0, wxEXPAND);
//  vboxMain->Add(-1, SEPARATOR);

    // End hboxSHA1Output

    hboxCompare = new wxBoxSizer(wxHORIZONTAL);

    lblCompare = new wxStaticText(pnlMain, wxID_ANY, _T("Compare Checksums:"));
    txtCompare = new wxTextCtrl(pnlMain, wxID_ANY);
    btnCompare = new wxButton(pnlMain, ID_COMPARE, _T("&Compare"));

    hboxCompare->Add(lblCompare, 0, wxALIGN_CENTER_VERTICAL | wxRIGHT, 5);
    hboxCompare->Add(txtCompare, 1, wxALIGN_CENTER_VERTICAL);
    hboxCompare->Add(btnCompare, 0, wxALIGN_CENTER_VERTICAL | wxLEFT, 5);

    vboxMain->Add(hboxCompare, 0, wxEXPAND);
    // Bottom hbox, no separator

    // End hboxCompare

    pnlMain->SetSizer(vboxMain);

    vboxMargin->Add(pnlMain, 1, wxEXPAND | wxALL, MARGIN);
    vboxMargin->SetSizeHints(this);
    pnlMargin->SetSizer(vboxMargin);

    // :TODO: 2012-05-27 gbchaosmaster - Have txtFile accept Enter.
    // It will process the file, but we'll need a check to make sure that it
    // exists before we go on.

    // Connect main window events
    Connect(ID_BROWSE, wxEVT_COMMAND_BUTTON_CLICKED,
            wxCommandEventHandler(frmMain::selectFile));
    Connect(ID_COPY_MD5, wxEVT_COMMAND_BUTTON_CLICKED,
            wxCommandEventHandler(frmMain::copyMD5));
//  Connect(ID_COPY_SHA1, wxEVT_COMMAND_BUTTON_CLICKED,
//          wxCommandEventHandler(frmMain::copySHA1));
    Connect(ID_COMPARE, wxEVT_COMMAND_BUTTON_CLICKED,
            wxCommandEventHandler(frmMain::compare));

    /** Configure Form **/

    // Set the minimum/startup width, height will be figured out by the sizers
    SetMinSize(wxSize(MIN_WIDTH, -1));
    // Don't let the user change the height
    SetMaxSize(wxSize(wxDefaultCoord, GetSize().y));
    // Center the window on screen on startup
    Center();
}

void frmMain::selectFile(wxCommandEvent &WXUNUSED(event))
{
    // Let the user select a file from the dialog
    wxFileDialog dlgOpenFile(this, _T("Select a File"), wxGetHomeDir(),
                             wxEmptyString, _T("All Files|*.*"),
                             wxFD_OPEN | wxFD_FILE_MUST_EXIST);

    // If they pressed cancel, stop
    if (dlgOpenFile.ShowModal() == wxID_CANCEL) return;

    // Set the selected file path to the text box
    txtFile->SetValue(dlgOpenFile.GetPath());

    // Read the entire file into a wxString
    wxString contents;
    wxFFile file(dlgOpenFile.GetPath());
    file.ReadAll(&contents);
    file.Close();

    txtMD5->SetValue(encryptMD5(contents));
}

void frmMain::compare(wxCommandEvent &WXUNUSED(event))
{
    // Trim whitespace from ends of the pasted hash
    txtCompare->SetValue(txtCompare->GetValue().Trim()); // End
    txtCompare->SetValue(txtCompare->GetValue().Trim(false)); // Beginning

    // Check if it's a good/bad match and tell them the result
    if (txtCompare->GetValue() == txtMD5->GetValue())
        wxMessageBox(_T("The MD5 hash matches."), _T("Good Match"));
    else wxMessageBox(_T("The MD5 hash does not match."), _T("Bad Match"));
}

void frmMain::copyChecksum(wxString str)
{
    if (!str.IsEmpty())
    {
        Clipboard::SetClipboard(str);

        if (Clipboard::GetClipboard() == str)
            wxMessageBox(_T("Checksum copied successfully."), _T("Success"));
    }
}
void frmMain::copyMD5(wxCommandEvent &WXUNUSED(event))
{
    copyChecksum(txtMD5->GetValue());
}
//void frmMain::copySHA1(wxCommandEvent &WXUNUSED(event))
//{
//  copyChecksum(txtSHA1->GetValue());
//}

/** Miscellaneous Functions **/

void frmMain::showAbout(wxCommandEvent &WXUNUSED(event))
{
    wxAboutDialogInfo info;
    info.SetName(_T("Checksum Finder"));
    info.SetVersion(_T("0.1"));
    info.SetDescription(_T("Finds the MD5/SHA-1 checksums of a file."));
    info.SetCopyright(_T("Copyright (c) 2012 Vinny Diehl"));

    wxAboutBox(info);
}

void frmMain::quit(wxCommandEvent &WXUNUSED(event))
{
    Close();
}

